File: 백준/Gold/1781. 컵라면/컵라면.py
import heapq
import sys
input = sys.stdin.readline
n = int(input())
cup = []
last_deadline = 0
for _ in range(n):
    a, b = map(int, input().split())
    last_deadline = max(last_deadline, a)
    cup.append((a, b))
cup.sort(key=lambda x:(-x[0], -x[1]))

que = []
idx, ans = 0, 0
while last_deadline > 0:
    while idx < n and cup[idx][0] >= last_deadline:
        heapq.heappush(que, -cup[idx][1])
        idx += 1
    last_deadline -= 1
    if que:
        ans -= heapq.heappop(que)
print(ans)

File: 백준/Gold/1781. 컵라면/컵라면.java
import java.io.*;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;
        int n = Integer.parseInt(br.readLine());

        int cup[][] = new int[n][2];
        for (int i=0; i<n; i++) {
            st = new StringTokenizer(br.readLine());
            cup[i][0] = Integer.parseInt(st.nextToken());
            cup[i][1] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(cup, (a, b) -> {
            if (a[0] == b[0]) return b[1] - a[1];
            return a[0] - b[0];
        });
        PriorityQueue<Integer> que = new PriorityQueue<>();

        for (int i=0; i<n; i++) {
            que.add(cup[i][1]);

            while (que.size() > cup[i][0]) {
                que.poll();
            }
        }
        int ans = 0;
        for (int s: que) {
            ans += s;
        }

        bw.write("" + ans);
        bw.flush();
        bw.close();
    }
}